pkg/controller/generic/cleanup/cleanup.go

```diff
@@ -0,0 +1,431 @@
+		return h.finalizerRemoval(ctx, rt, l, input)
+	}
+
+	return nil
+}
+
+// Inputs implements controller.Controller interface.
+func (h *handler[I, O]) Inputs() []controller.Input {
+	if h.inputs != nil {
+		return h.inputs()
+	}
+
+	return nil
+}
+
+// Outputs implements controller.Controller interface.
+func (h *handler[I, O]) Outputs() []controller.Output {
+	if h.outputs != nil {
+		return h.outputs()
+	}
+
+	return nil
+}
+
+// HasNoOutputs is a helper function to create a [Handler] that waits for all outputs to be removed.
+func HasNoOutputs[O generic.ResourceWithRD, I generic.ResourceWithRD](
+	listOptions func(I) state.ListOption,
+) Handler[I] {
+	return &handler[I, O]{
+		finalizerRemoval: func(ctx context.Context, r controller.Runtime, logger *zap.Logger, input I) error {
+			var (
+				zeroOutput          O
+				zeroOutputType      = zeroOutput.ResourceDefinition().Type
+				zeroOutputNamespace = zeroOutput.ResourceDefinition().DefaultNamespace
+			)
+
+			list, err := r.List(
+				ctx,
+				resource.NewMetadata(zeroOutputNamespace, zeroOutputType, "", resource.VersionUndefined),
+				listOptions(input),
+			)
+			if err != nil {
+				return fmt.Errorf("error listing resources on input %q: %w", input.Metadata().ID(), err)
+			}
+
+			if len(list.Items) > 0 {
+				logger.Info(
+					"waiting for resources to be destroyed",
+					zap.String("resource", zeroOutputType),
+					zap.Int("count", len(list.Items)),
+				)
+
+				return xerrors.NewTagged[SkipReconcileTag](errors.New("waiting for resources to be destroyed"))
+			}
+
+			return nil
+		},
+		inputs: func() []controller.Input {
+			var zeroOutput O
+
+			return []controller.Input{{
+				Namespace: zeroOutput.ResourceDefinition().DefaultNamespace,
+				Type:      zeroOutput.ResourceDefinition().Type,
+				Kind:      controller.InputWeak,
+			}}
+		},
+	}
+}
+
+// RemoveOutputs is a helper function to create a [Handler] that removes all outputs on input teardown. It ignores
+// resource ownership using [controller.IgnoreOwner].
+func RemoveOutputs[O generic.ResourceWithRD, I generic.ResourceWithRD](
+	listOptions func(I) state.ListOption,
+) Handler[I] {
+	return &handler[I, O]{
+		finalizerRemoval: func(ctx context.Context, r controller.Runtime, logger *zap.Logger, input I) error {
+			var (
+				zeroOutput          O
+				zeroOutputType      = zeroOutput.ResourceDefinition().Type
+				zeroOutputNamespace = zeroOutput.ResourceDefinition().DefaultNamespace
+			)
+
+			list, err := safe.ReaderList[O](
+				ctx,
+				r,
+				resource.NewMetadata(zeroOutputNamespace, zeroOutputType, "", resource.VersionUndefined),
+				listOptions(input),
+			)
+			if err != nil {
+				return fmt.Errorf("error listing resources on input %q: %w", input.Metadata().ID(), err)
+			}
+
+			var multiErr error
+
+			var inTearDown int
+
+			for iter := safe.IteratorFromList(list); iter.Next(); {
+				out := iter.Value()
+
+				ready, err := r.Teardown(ctx, out.Metadata(), controller.WithOwner(""))
```

Comment:
should we skip here resources which have `owner != ""` ? i.e. those resources which should be cleaned up by other controllers?

we should probably still wait on them, but we shouldn't attempt tearing them down, as this will fail?